File: warez/datamodel/src/main/scala/warez/KeyedSchemas.scala
package warez

import pipelines.streamlets._
import pipelines.streamlets.avro._

class KeyedSchemas {
  implicit val productKeyed: Keyed[Product] = new Keyed[Product] { def key(p: Product) = p.id.toString }
  implicit val productCodec: KeyedSchema[Product] = AvroKeyedSchema[Product](Product.SCHEMA$)

  implicit val stockUpdateKeyed: Keyed[StockUpdate] = new Keyed[StockUpdate] { def key(p: StockUpdate) = p.productId.toString }
  implicit val stockUpdateCodec: KeyedSchema[StockUpdate] = AvroKeyedSchema[StockUpdate](StockUpdate.SCHEMA$)

  implicit val priceUpdateKeyed: Keyed[PriceUpdate] = new Keyed[PriceUpdate] { def key(p: PriceUpdate) = p.productId.toString }
  implicit val priceUpdateCodec: KeyedSchema[PriceUpdate] = AvroKeyedSchema[PriceUpdate](PriceUpdate.SCHEMA$)

  implicit val recommenderModelKeyed: Keyed[RecommenderModel] = new Keyed[RecommenderModel] { def key(m: RecommenderModel) = m.modelId.toString }
  implicit val recommenderModelCodec: KeyedSchema[RecommenderModel] = AvroKeyedSchema[RecommenderModel](RecommenderModel.SCHEMA$)
}

object KeyedSchemas extends KeyedSchemas {
  def instance: KeyedSchemas = this
}

File: sensor-data-java/src/main/java/pipelines/examples/sensordata/KeyedSchemas.java
package pipelines.examples.sensordata;

import org.apache.avro.Schema;
import pipelines.streamlets.*;
import pipelines.streamlets.avro.*;


public final class KeyedSchemas {
  static final AvroKeyedSchema<Metric> metricKeyed = 
    new AvroKeyedSchema<Metric>() {
      @Override public Schema schema() { return Metric.SCHEMA$; }
      @Override public Keyed<Metric> keyed() { 
	      return new Keyed<Metric>() { 
          @Override public String key(Metric m) {
            return m.getDeviceId().toString() + m.getTimestamp().toString(); 
          }
        };
      }
    };

  static final AvroKeyedSchema<InvalidMetric> invalidMetricKeyed = 
    new AvroKeyedSchema<InvalidMetric>() {
      @Override public Schema schema() { return InvalidMetric.SCHEMA$; }
      @Override public Keyed<InvalidMetric> keyed() { 
	      return new Keyed<InvalidMetric>() { 
          @Override public String key(InvalidMetric m) {
            return m.getMetric().toString();
    	    }
    	  };
      }
    };

  static final AvroKeyedSchema<SensorData> sensorDataKeyed = 
    new AvroKeyedSchema<SensorData>() {
      @Override public Schema schema() { return SensorData.SCHEMA$; }
      @Override public Keyed<SensorData> keyed() { 
      	return new Keyed<SensorData>() { 
          @Override public String key(SensorData s) {
            return s.getDeviceId().toString() + s.getTimestamp().toString();
          }
        };
      }
    };

  public static final KeyedSchemas instance = new KeyedSchemas();
}
